plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,216 @@
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Simple abstract class that represent record stored in the Query Insight Framework
+ *
+ * @param <T> The value type associated with the record
+ * @opensearch.internal
+ */
+public abstract class SearchQueryRecord<T extends Number & Comparable<T>>
+    implements
+        Comparable<SearchQueryRecord<T>>,
+        Writeable,
+        ToXContentObject {
+
+    private static final Logger log = LogManager.getLogger(SearchQueryRecord.class);
+    protected static final String TIMESTAMP = "timestamp";
+    protected static final String SEARCH_TYPE = "searchType";
+    protected static final String SOURCE = "source";
+    protected static final String TOTAL_SHARDS = "totalShards";
+    protected static final String INDICES = "indices";
+    protected static final String PROPERTY_MAP = "propertyMap";
+    protected static final String VALUE = "value";
+
+    private final Long timestamp;
+
+    private final SearchType searchType;
+
+    private final String source;
+
+    private final int totalShards;
+
+    private final String[] indices;
+
+    // TODO: add user-account which initialized the request in the future
+    private final Map<String, Object> propertyMap;
+
+    private T value;
+
+    public SearchQueryRecord(final StreamInput in) throws IOException, ClassCastException {
+        this.timestamp = in.readLong();
+        this.searchType = SearchType.fromString(in.readString().toLowerCase(Locale.ROOT));
+        this.source = in.readString();
+        this.totalShards = in.readInt();
+        this.indices = in.readStringArray();
+        this.propertyMap = in.readMap();
+    }
+
+    public SearchQueryRecord(
+        final Long timestamp,
+        final SearchType searchType,
+        final String source,
+        final int totalShards,
+        final String[] indices,
+        final Map<String, Object> propertyMap,
+        final T value
+    ) {
+        this(timestamp, searchType, source, totalShards, indices, propertyMap);
+        this.value = value;
+    }
+
+    public SearchQueryRecord(
+        final Long timestamp,
+        final SearchType searchType,
+        final String source,
+        final int totalShards,
+        final String[] indices,
+        final Map<String, Object> propertyMap
+    ) {
+        this.timestamp = timestamp;
+        this.searchType = searchType;
+        this.source = source;
+        this.totalShards = totalShards;
+        this.indices = indices;
+        this.propertyMap = propertyMap;
+    }
+
+    /**
+     * The timestamp of the top query.
+     */
+    public Long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * The manner at which the search operation is executed.
+     */
+    public SearchType getSearchType() {
+        return searchType;
+    }
+
+    /**
+     * The search source that was executed by the query.
+     */
+    public String getSource() {
+        return source;
+    }
```

Comment:
These getters should not be public. Also, we might not need these if we leverage superclass takes care of its own ser/deser

Reply:
We might not need them right now but in the future they might be required for other query insights features, for example, adding customized processors. Those processors may need to use those getters to get certain value of the records.

Reply:
Let us add these as and when needed? Otherwise, we have added dead code without any code coverage.